src/react/utils.js

```diff
@@ -962,40 +984,116 @@ function applyClonedTemporalAlias(realm: Realm, props: ObjectValue, clonedProps:
     { skipInvariant: true }
   );
   invariant(temporalTo instanceof AbstractObjectValue);
-  invariant(clonedProps instanceof ObjectValue);
-  temporalTo.values = new ValuesDomain(clonedProps);
-  clonedProps.temporalAlias = temporalTo;
+  invariant(clonedObj instanceof ObjectValue);
+  temporalTo.values = new ValuesDomain(clonedObj);
+  clonedObj.temporalAlias = temporalTo;
   // Store the args for the temporal so we can easily clone
   // and reconstruct the temporal at another point, rather than
   // mutate the existing temporal
   realm.temporalAliasArgs.set(temporalTo, newTemporalArgs);
 }
 
-export function cloneProps(realm: Realm, props: ObjectValue, newChildren?: Value): ObjectValue {
-  let clonedProps = new ObjectValue(realm, realm.intrinsics.ObjectPrototype);
-
-  for (let [propName, binding] of props.properties) {
-    if (binding && binding.descriptor && binding.descriptor.enumerable) {
-      if (newChildren !== undefined && propName === "children") {
-        Properties.Set(realm, clonedProps, propName, newChildren, true);
-      } else {
-        Properties.Set(realm, clonedProps, propName, getProperty(realm, props, propName), true);
+export function cloneObject(
+  realm: Realm,
+  obj: ObjectValue | AbstractObjectValue,
+  cloneTemporalAlias: boolean,
+  alreadyCloned?: Map<ObjectValue | AbstractObjectValue, ObjectValue | AbstractObjectValue> = new Map()
+): ObjectValue | AbstractObjectValue {
+  if (alreadyCloned.has(obj)) {
+    let _obj = alreadyCloned.get(obj);
+    invariant(_obj instanceof ObjectValue || _obj instanceof AbstractObjectValue);
+    return _obj;
+  }
+  if (obj instanceof ObjectValue) {
+    let clonedObj = new ObjectValue(realm, realm.intrinsics.ObjectPrototype);
+    alreadyCloned.set(obj, clonedObj);
+    let isFinalObject = obj.mightBeFinalObject();
```

Comment:
This doesn't look right. Just because it *might be* final doesn't actually *make* it final. The result has a stronger assertion that might be false for the original object. Can we preserve the level of knowledge without distorting it?